systems/hub/hub/pkg/server/router.go

```diff
@@ -0,0 +1,290 @@
+	capps.GET("/", nil, tonic.Handler(r.listAllAppsHandler, http.StatusOK))
+	capps.GET("/location", nil, tonic.Handler(r.cappLocationHandler, http.StatusOK))
+}
+
+func (r *Router) cappGetHandler(c *gin.Context, req *CAppRequest) error {
+	log.Infof("Getting artifact: %s %s", req.Name, req.ArtifactName)
+
+	v, ext, err := parseArtifactName(req.ArtifactName)
+	if err != nil {
+		return rest.HttpError{
+			HttpCode: http.StatusNotFound,
+			Message:  "Artifact name is not valid",
+		}
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), r.storageRequestTimeout)
+	defer cancel()
+
+	rd, err := r.storage.GetFile(ctx, req.Name, v, ext)
+	if err != nil {
+		return err
+	}
+	defer rd.Close()
+
+	_, err = io.Copy(c.Writer, rd)
+	if err != nil {
+		if minio.ToErrorResponse(err).Code == "NoSuchKey" {
+			return rest.HttpError{
+				HttpCode: http.StatusNotFound,
+				Message:  "Artifact not found",
+			}
+		}
+	}
+
+	c.Header("Content-Type", "application/octet-stream")
+	c.Header("Content-Disposition",
+		fmt.Sprintf("attachment; filename=%s-%s%s", req.Name, v.String(), ext))
+
+	return nil
+}
+
+func parseArtifactName(name string) (ver *semver.Version, ext string, err error) {
+	if strings.HasSuffix(name, pkg.TarGzExtension) {
+		name = strings.TrimSuffix(name, pkg.TarGzExtension)
+		ext = pkg.TarGzExtension
+	} else if strings.HasSuffix(name, pkg.ChunkIndexExtension) {
+		name = strings.TrimSuffix(name, pkg.ChunkIndexExtension)
+		ext = pkg.ChunkIndexExtension
+	} else {
+		return nil, "", fmt.Errorf("Unsupported extension")
+	}
+
+	ver, err = semver.NewVersion(name)
+	if err != nil {
+		return nil, "", errors.Wrap(err, "failed to parse version")
+	}
+
+	return ver, ext, nil
+}
+
+func (r *Router) cappPutHandler(c *gin.Context) error {
+	name := c.Param("name")
+	ver := c.Param("version")
+
+	log.Infof("Adding artifact: %s %s", name, ver)
+
+	ctx, cancel := context.WithTimeout(context.Background(), r.storageRequestTimeout)
+	defer cancel()
+
+	v, err := r.parseVersion(ver)
+	if err != nil {
+		return err
+	}
+
+	bufReader := NewBufReader(c.Request.Body)
+	defer c.Request.Body.Close()
+
+	uncompressedStream, err := gzip.NewReader(bufReader)
+	if err != nil {
+		log.Infof("Failed to read gz file: %v", err)
+
+		return rest.HttpError{
+			HttpCode: http.StatusBadRequest,
+			Message:  "Not a tar.gz file",
+		}
+	}
+
+	tr := tar.NewReader(uncompressedStream)
+
+	_, err = tr.Next()
+	if err != nil {
+		log.Infof("Failed to read tar file: %v", err)
+
+		return rest.HttpError{
+			HttpCode: http.StatusBadRequest,
+			Message:  "Not a tar.gz file",
+		}
+	}
+
+	bufReader.Reset()
+
+	loc, err := r.storage.PutFile(ctx, name, v, pkg.TarGzExtension, bufReader)
+	if err != nil {
+		log.Errorf("Error adding artifact: %s %s", name, ver)
+
+		return err
+	}
+
+	go func() {
+		err = r.chunker.Chunk(name, v, loc)
+		if err != nil {
+			log.Errorf("Error chunking artifact: %s %s. Error: %+v", name, ver, err)
+		}
+	}()
+
+	capp := &epb.CappCreatedEvent{
```

Comment:
this doesn't mean you have capp ready. (It's just a request). May be right place for this is in distributor after chunking is done.